### PR TITLE
Added Fix - Allow nxos_l2_interfaces to append the allowed vlans list

### LINKED ISSUE #65332: Add functionality to `nxos_l2_interfaces`  to append the allowed vlans list
##### SUMMARY

Currently, the `nxos_l2_interfaces` module does not provide an option to append vlans to that allowed vlan list. When using the `state: merged`  any config that was untouched by the task remains but any configuration provided will replace the current config. This option was available in the `nxos_l2_interface` with the `trunk_vlans` parm.  The module is currently sending the `switchport trunk allowed vlan <#>` and not `switchport trunk allowed vlan add <#>` so a full list of allowed vlans must be supplied. 

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

nxos_l2_interfaces
##### ANSIBLE VERSION

```paste below
ansible 2.9.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/Users/bdudas/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.9.1/libexec/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.5 (default, Nov  1 2019, 02:16:32) [Clang 11.0.0 (clang-1100.0.33.8
```
##### OS / ENVIRONMENT

```
an-nxos-02# show version
Cisco Nexus Operating System (NX-OS) Software
TAC support: http://www.cisco.com/tac
Documents: http://www.cisco.com/en/US/products/ps9372/tsd_products_support_series_home.html
Copyright (c) 2002-2016, Cisco Systems, Inc. All rights reserved.
The copyrights to certain works contained herein are owned by
other third parties and are used and distributed under license.
Some parts of this software are covered under the GNU Public
License. A copy of the license is available at
http://www.gnu.org/licenses/gpl.html.

NX-OSv is a demo version of the Nexus Operating System

Software
  loader:    version N/A
  kickstart: version 7.3(0)D1(1)
  system:    version 7.3(0)D1(1)
  kickstart image file is: bootflash:///titanium-d1-kickstart.7.3.0.D1.1.bin
  kickstart compile time:  1/11/2016 16:00:00 [02/11/2016 10:30:12]
  system image file is:    bootflash:///titanium-d1.7.3.0.D1.1.bin
  system compile time:     1/11/2016 16:00:00 [02/11/2016 13:08:11]


Hardware
  cisco NX-OSv Chassis ("NX-OSv Supervisor Module")
  QEMU Virtual CPU version 2.5 with 3064740 kB of memory.
  Processor Board ID TM000B0000B

  Device name: an-nxos-02
  bootflash:    3184776 kB

Kernel uptime is 0 day(s), 2 hour(s), 44 minute(s), 51 second(s)


plugin
  Core Plugin, Ethernet Plugin

Active Package(s)
```

##### STEPS TO REPRODUCE

```yaml
---
- hosts: nxos
  gather_facts: false
  vars:
    new_vlans: 200

  tasks:
    - name: add vlans to allowed list
      nxos_l2_interfaces:
        config:
          - name: Ethernet2/31
            trunk:
              allowed_vlans: "{{ new_vlans }}"
        state: merged
...
```

##### RESULTS

Starting Interface Config:
```
an-nxos-02# show run interface Ethernet 2/31

!Command: show running-config interface Ethernet2/31
!Time: Wed Nov 27 15:57:00 2019

version 7.3(0)D1(1)

interface Ethernet2/31
  shutdown
  switchport
  switchport trunk native vlan 99
  switchport trunk allowed vlan 101,104-106
```
Post Job interface Config:
```
an-nxos-02# show run interface Ethernet 2/31

!Command: show running-config interface Ethernet2/31
!Time: Wed Nov 27 16:04:53 2019

version 7.3(0)D1(1)

interface Ethernet2/31
  shutdown
  switchport
  switchport trunk native vlan 99
  switchport trunk allowed vlan 200
```

```paste below
ansible-playbook 2.9.1
  config file = /Users/fubar/dev-net/nxos/ansible.cfg
  configured module search path = ['/Users/fubar/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.9.1/libexec/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.7.5 (default, Nov  1 2019, 02:16:32) [Clang 11.0.0 (clang-1100.0.33.8)]
Using /Users/fubar/dev-net/nxos/ansible.cfg as config file
host_list declined parsing /Users/fubar/dev-net/inventory/hosts as it did not pass its verify_file() method
script declined parsing /Users/fubar/dev-net/inventory/hosts as it did not pass its verify_file() method
auto declined parsing /Users/fubar/dev-net/inventory/hosts as it did not pass its verify_file() method
Parsed /Users/fubar/dev-net/inventory/hosts inventory source with ini plugin
 _____________________________
< PLAYBOOK: allowed_vlans.yml >
 -----------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

1 plays in allowed_vlans.yml
 _____________
< PLAY [nxos] >
 -------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

META: ran handlers
 __________________________________
< TASK [add vlans to allowed list] >
 ----------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /Users/fubar/dev-net/nxos/allowed_vlans.yml:8
<nxos-01> ESTABLISH LOCAL CONNECTION FOR USER: fubar
<nxos-01> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828 `" && echo ansible-tmp-1574871651.6025379-270697238693828="` echo /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828 `" ) && sleep 0'
<nxos-01> Attempting python interpreter discovery
<nxos-01> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<nxos-01> Python interpreter discovery fallback (unsupported platform for extended discovery: darwin)
Using module file /usr/local/Cellar/ansible/2.9.1/libexec/lib/python3.7/site-packages/ansible/modules/network/nxos/nxos_l2_interfaces.py
<nxos-01> PUT /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/tmpjicqhgg0 TO /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828/AnsiballZ_nxos_l2_interfaces.py
<nxos-01> EXEC /bin/sh -c 'chmod u+x /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828/ /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828/AnsiballZ_nxos_l2_interfaces.py && sleep 0'
<nxos-01> EXEC /bin/sh -c '/usr/bin/python /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828/AnsiballZ_nxos_l2_interfaces.py && sleep 0'
<nxos-01> EXEC /bin/sh -c 'rm -f -r /Users/fubar/.ansible/tmp/ansible-local-351511mqamcek/ansible-tmp-1574871651.6025379-270697238693828/ > /dev/null 2>&1 && sleep 0'
[WARNING]: The value 200 (type int) in a string field was converted to u'200' (type string). If this does not look like
what you expect, quote the entire value to ensure it does not change.

[WARNING]: Platform darwin on host nxos-01 is using the discovered Python interpreter at
/usr/bin/python, but future installation of another Python interpreter could change this. See
https://docs.ansible.com/ansible/2.9/reference_appendices/interpreter_discovery.html for more information.

changed: [nxos-01] => {
    "after": [
        {
            "name": "Ethernet2/31",
            "trunk": {
                "allowed_vlans": "200",
                "native_vlan": 99
            }
        }
    ],
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "before": [
        {
            "name": "Ethernet2/31",
            "trunk": {
                "allowed_vlans": "101,104,105,106",
                "native_vlan": 99
            }
        }
    ],
    "changed": true,
    "commands": [
        "interface Ethernet2/31",
        "switchport trunk allowed vlan 200"
    ],
    "invocation": {
        "module_args": {
            "config": [
                {
                    "access": null,
                    "name": "Ethernet2/31",
                    "trunk": {
                        "allowed_vlans": "200",
                        "native_vlan": null
                    }
                }
            ],
            "state": "merged"
        }
    }
}
META: ran handlers
META: ran handlers
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

nxos-01 : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### PR DESCRIPTION
**Pull request by GomathiselviS**
##### SUMMARY
Fixed #65332

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
l2_interfaces.py

### CODE REVIEW

File: lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py [hunk 1]
```diff
@@ -255,7 +255,10 @@ def add_commands(self, d):
         if 'vlan' in d:
             commands.append(cmd + 'access vlan ' + str(d['vlan']))
         if 'allowed_vlans' in d:
-            commands.append(cmd + 'trunk allowed vlan ' + str(d['allowed_vlans']))
+            if vlan_exists:
```
Comment (justjais) [inline lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: @GomathiselviS code changes LGTM, can u add on 1 test covering the use case.

**Review by justjais — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py (hunk 1 above)
Comment (GomathiselviS) [reply lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: Added testcase

**Review by justjais — APPROVED** [review]

File: lib/ansible/modules/network/eos/eos_eapi.py [hunk 2]
```diff
@@ -268,7 +268,6 @@ def add(cmd):
             add('no protocol unix-socket')
         else:
             add('protocol unix-socket')
-
```
Comment (pabelanger) [inline lib/ansible/modules/network/eos/eos_eapi.py]: This looks unrelated, if you want to revert.

**Comment by NilashishC** [comment]
@chrisvanheuveln @mikewiebe could you have a look at this PR? Thanks!

**Comment by mikewiebe** [comment]
> @chrisvanheuveln @mikewiebe could you have a look at this PR? Thanks!

Will do.

**Review by NilashishC — COMMENTED** [review]
Why are we touching `lib/ansible/module_utils/network/common/cfg/base.py`? Is this a rebase spillover?

**Comment by pabelanger** [comment]
Agree with @NilashishC if would be helpful to understand change to cfg/base.py, as it will affect all platforms

**Comment by GomathiselviS** [comment]
@NilashishC  @Qalthos  yeah .. I had to do this to resolve merge conflicts.

**Comment by GomathiselviS** [comment]
recheck

**Comment by GomathiselviS** [comment]
recheck

**Comment by mikewiebe** [comment]
I believe you should add some documentation here (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/network/nxos/nxos_l2_interfaces.py#L78) to indicate that this parameter will add vlans to the existing list when `state` is `merged` and vlans already exist.

File: lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py (hunk 1 above)
Comment (mikewiebe) [inline lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: I just ran a test and discovered that adding vlans using a range to an already existing list is not idempotent.  Logic should be added to expand a range if provided to match what is in the running config.

First I applied this playbook:

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 10
              allowed_vlans: 35-40 
            #access:
            #  vlan: 30
        state:  merged
```

Re-running this playbook is idempotent.

I then tried to add a few more vlans using the following playbook.

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 10
              allowed_vlans: 50-60 
            #access:
            #  vlan: 30
        state:  merged

```

It adds them but it's not idempotent.

```shell
    "after": [
        {
            "name": "Ethernet1/3", 
            "trunk": {
                "allowed_vlans": "35-40,50-60", 
                "native_vlan": 10
            }
        }
    ], 
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    }, 
    "before": [
        {
            "name": "Ethernet1/3", 
            "trunk": {
                "allowed_vlans": "35,36,37,38,39,40,50,51,52,53,54,55,56,57,58,59,60", 
                "native_vlan": 10
            }
        }
    ], 
    "changed": true, 
    "commands": [
        "interface Ethernet1/3", 
        "switchport trunk allowed vlan add 50,51,52,53,54,55,56,57,58,59,60"
    ], 
    "invocation": {
        "module_args": {
            "config": [
                {
                    "access": null, 
                    "name": "Ethernet1/3", 
                    "trunk": {
                        "allowed_vlans": "50,51,52,53,54,55,56,57,58,59,60", 
                        "native_vlan": 10
                    }
                }
            ], 
            "state": "merged"
        }
    }
}
```

If the state is merged we should only send commands to add the new vlans if they are not already part of the list.  If I include the entire list of configured vlans it's idempotent but it should be idempotent if I only provide a partial list but that list is already configured.

Comment (mikewiebe) [reply lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: **Another problem** 

The `replaced` state does not appear to work properly.

I start with the following config:

```shell
interface Ethernet1/3
  switchport
  switchport trunk native vlan 40
  switchport trunk allowed vlan 35-40,50-60
```

I run the following playbook:

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 50
              allowed_vlans: 45-70
        state: replaced
```

Here is my config after running the playbook.

```shell
interface Ethernet1/3
  switchport
  switchport trunk native vlan 50
  switchport trunk allowed vlan 35-40,45-70
```

The `replaced` state should replace my running config with whatever is defined in the playbook but as you can see it left behind 35-40.

Comment (mikewiebe) [reply lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: **NOTE: ** Same problem with `state: overridden`

Start with the following configs:

```
interface Ethernet1/3
  switchport
  switchport trunk native vlan 50
  switchport trunk allowed vlan 35-40,45-70


interface Ethernet1/4
  switchport
  switchport trunk native vlan 60
  switchport trunk allowed vlan 35-40

```

Run the following playbook

```yaml
    - name: Merge with existing vlans
      nxos_l2_interfaces:
        config:
          - name: "Ethernet1/3"
            trunk:
              native_vlan: 50
              allowed_vlans: 45-70 
        state: overridden 
```

State after running the playbook:

```
interface Ethernet1/3
  switchport
  switchport trunk allowed vlan 35-40,45-70

interface Ethernet1/4
  switchport

```

State of Ethernet1/4 is correct but like state replaced Ethernet1/3 should only have vlans 45-70

**Review by mikewiebe — CHANGES_REQUESTED** [review]
Ran several tests with the various states and found some issues.  Please see additional comments for details.

**Comment by GomathiselviS** [comment]
@mikewiebe I will look in to this.

File: lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py (hunk 1 above)
Comment (GomathiselviS) [reply lib/ansible/module_utils/network/nxos/config/l2_interfaces/l2_interfaces.py:258]: @mikewiebe : I have made few changes. I tested with various combinations. Can you please take a look? I will work on adding necessary testcases and documentation.

**Comment by GomathiselviS** [comment]
recheck

**Comment by NilashishC** [comment]
recheck

**Comment by mikewiebe** [comment]
> @mikewiebe Can you please have a look?

Yes, I will look at this today.

**Comment by GomathiselviS** [comment]
> > @mikewiebe Can you please have a look?
> 
> Yes, I will look at this today but on a quick scan I don't see any new integration tests that cover the scenarios I described in my comments. Can you please add at least one test for `merged`, `replaced`, and `overridden` states? Want to make sure we have tests for the functionality so it does not get broken by future commits

@mikewiebe : I have added the necessary testcases. You can check now. Let me know if I have missed any of the scenarios.

**Comment by NilashishC** [comment]
recheck
